server/session/handler_item_stack_request.go

```diff
@@ -183,6 +190,128 @@ func call(ctx *event.Context, slot int, it item.Stack, f func(ctx *event.Context
+	if !ok {
+		return fmt.Errorf("recipe with network id %v does not exist", a.RecipeNetworkID)
+	}
+	_, shaped := craft.(recipe.ShapedRecipe)
+	_, shapeless := craft.(recipe.ShapelessRecipe)
+	if !shaped && !shapeless {
+		return fmt.Errorf("recipe with network id %v is not a shaped or shapeless recipe", a.RecipeNetworkID)
+	}
+
+	input := craft.Input()
+	size := s.craftingSize()
+	offset := s.craftingOffset()
+	consumed := make([]uint32, 0, size)
+	for _, expected := range input {
+		var processed bool
+		for slot := offset; slot < offset+size; slot++ {
+			if slices.Contains(consumed, slot) {
+				// We've already consumed this slot, skip it.
+				continue
+			}
+			has, _ := s.ui.Item(int(slot))
+			if has.Empty() != expected.Empty() || has.Count() < expected.Count() || !has.Comparable(expected) {
+				// We can't process this item, as it's not a part of the recipe.
+				continue
+			}
+			processed, consumed = true, append(consumed, slot)
+			st := has.Grow(-expected.Count())
+			h.setItemInSlot(protocol.StackRequestSlotInfo{
+				ContainerID:    containerCraftingGrid,
+				Slot:           byte(slot),
+				StackNetworkID: item_id(st),
+			}, st, s)
+			break
+		}
+		if !processed {
+			return fmt.Errorf("recipe %v: could not consume expected item: %v", a.RecipeNetworkID, expected)
+		}
+	}
+
+	output := craft.Output()
+	h.setItemInSlot(protocol.StackRequestSlotInfo{
+		ContainerID:    containerCraftingResult,
+		Slot:           craftingResultIndex,
+		StackNetworkID: item_id(output[0]),
+	}, output[0], s)
+	return nil
+}
+
+// handleAutoCraft handles the AutoCraftRecipe request action.
+func (h *ItemStackRequestHandler) handleAutoCraft(a *protocol.AutoCraftRecipeStackRequestAction, s *Session) error {
+	craft, ok := s.recipes[a.RecipeNetworkID]
+	if !ok {
+		return fmt.Errorf("recipe with network id %v does not exist", a.RecipeNetworkID)
+	}
+	_, shaped := craft.(recipe.ShapedRecipe)
+	_, shapeless := craft.(recipe.ShapelessRecipe)
+	if !shaped && !shapeless {
+		return fmt.Errorf("recipe with network id %v is not a shaped or shapeless recipe", a.RecipeNetworkID)
+	}
+
+	input := make([]item.Stack, 0, len(craft.Input()))
+	for _, i := range craft.Input() {
+		input = append(input, i.Grow(i.Count()*(int(a.TimesCrafted)-1)))
+	}
+
+	expectancies := make([]item.Stack, 0, len(input))
+	for _, i := range input {
+		if i.Empty() {
+			// We don't actually need this item - it's empty, so avoid putting it in our expectancies.
+			continue
+		}
+
+		if ind := slices.IndexFunc(expectancies, func(stack item.Stack) bool {
+			return stack.Comparable(i)
+		}); ind >= 0 {
+			i = i.Grow(expectancies[ind].Count())
+			expectancies = slices.Delete(expectancies, ind, ind+1)
+		}
+		expectancies = append(expectancies, i)
+	}
+
+	for _, expected := range expectancies {
+		for id, inv := range map[byte]*inventory.Inventory{containerCraftingGrid: s.ui, containerFullInventory: s.inv} {
+			for slot, has := range inv.Slots() {
```

Comment:
If you use `inv.Items()` here, you can get rid of the `has.Empty()` check.

Reply:
I need the slot, and if I use Items, the slot won't be exact.

Reply:
Ah, right, might be better to reorder these loops though

Reply:
That'd cause mismatch issues if the items are split up between different inventories and vary